Mathlib/Algebra/Algebra/Basic.lean

```diff
@@ -0,0 +1,921 @@
+
+See the implementation notes in this file for discussion of the details of this definition.
+-/
+-- porting note: unsupported @[nolint has_nonempty_instance]
+class Algebra (R : Type u) (A : Type v) [CommSemiring R] [Semiring A] extends SMul R A,
+  R →+* A where
+  commutes' : ∀ r x, toRingHom r * x = x * toRingHom r
+  smul_def' : ∀ r x, r • x = toRingHom r * x
+#align algebra Algebra
+
+end Prio
+
+/-- Embedding `R →+* A` given by `algebra` structure. -/
+def algebraMap (R : Type u) (A : Type v) [CommSemiring R] [Semiring A] [Algebra R A] : R →+* A :=
+  Algebra.toRingHom
+#align algebra_map algebraMap
+
+/-- Coercion from a commutative semiring to an algebra over this semiring. -/
+@[coe] def Algebra.cast {R A : Type _} [CommSemiring R] [Semiring A] [Algebra R A] : R → A :=
+  algebraMap R A
+
+namespace algebraMap
+
+instance coeHTCT (R A : Type _) [CommSemiring R] [Semiring A] [Algebra R A] : CoeHTCT R A :=
+  ⟨Algebra.cast⟩
+#align algebra_map.has_lift_t algebraMap.coeHTCT
+
+section CommSemiringSemiring
+
+variable {R A : Type _} [CommSemiring R] [Semiring A] [Algebra R A]
+
+@[simp, norm_cast]
+theorem coe_zero : (↑(0 : R) : A) = 0 :=
+  map_zero (algebraMap R A)
+#align algebra_map.coe_zero algebraMap.coe_zero
+
+@[simp, norm_cast]
+theorem coe_one : (↑(1 : R) : A) = 1 :=
+  map_one (algebraMap R A)
+#align algebra_map.coe_one algebraMap.coe_one
+
+@[norm_cast]
+theorem coe_add (a b : R) : (↑(a + b : R) : A) = ↑a + ↑b :=
+  map_add (algebraMap R A) a b
+#align algebra_map.coe_add algebraMap.coe_add
+
+@[norm_cast]
+theorem coe_mul (a b : R) : (↑(a * b : R) : A) = ↑a * ↑b :=
+  map_mul (algebraMap R A) a b
+#align algebra_map.coe_mul algebraMap.coe_mul
+
+@[norm_cast]
+theorem coe_pow (a : R) (n : ℕ) : (↑(a ^ n : R) : A) = (a : A) ^ n :=
+  map_pow (algebraMap R A) _ _
+#align algebra_map.coe_pow algebraMap.coe_pow
+
+end CommSemiringSemiring
+
+section CommRingRing
+
+variable {R A : Type _} [CommRing R] [Ring A] [Algebra R A]
+
+@[norm_cast]
+theorem coe_neg (x : R) : (↑(-x : R) : A) = -↑x :=
+  map_neg (algebraMap R A) x
+#align algebra_map.coe_neg algebraMap.coe_neg
+
+end CommRingRing
+
+section CommSemiringCommSemiring
+
+variable {R A : Type _} [CommSemiring R] [CommSemiring A] [Algebra R A]
+
+open BigOperators
+
+-- direct to_additive fails because of some mix-up with polynomials
+@[norm_cast]
+theorem coe_prod {ι : Type _} {s : Finset ι} (a : ι → R) :
+    (↑(∏ i : ι in s, a i : R) : A) = ∏ i : ι in s, (↑(a i) : A) :=
+  map_prod (algebraMap R A) a s
+#align algebra_map.coe_prod algebraMap.coe_prod
+
+-- to_additive fails for some reason
+@[norm_cast]
+theorem coe_sum {ι : Type _} {s : Finset ι} (a : ι → R) :
+    ↑(∑ i : ι in s, a i) = ∑ i : ι in s, (↑(a i) : A) :=
+  map_sum (algebraMap R A) a s
+#align algebra_map.coe_sum algebraMap.coe_sum
+
+-- porting note: removed attribute [to_additive] coe_prod; why should this be a `to_additive`?
```

Comment:
Why should it not be?

Reply:
Because the additive instance is not `to_additive` of a multiplicative instance?